rfcs/0135-client-env-vars.md

```diff
@@ -0,0 +1,91 @@
+# RFC 0135 - Clients and Environment Variables
+* Comments: [#135](https://api.github.com/repos/taskcluster/taskcluster-rfcs/pulls/135)
+* Proposed by: @djmitche
+
+# Summary
+
+Client libraries should not automatically read environment variables.
+
+## Motivation
+
+Some client libraries currently automatically use environment variables
+ * `TASKCLUSTER_CLIENT_ID`
+ * `TASKCLUSTER_ACCESS_TOKEN`
+ * `TASKCLUSTER_CERTIFICATE`
+ * `TASKCLUSTER_ROOT_URL`
+if they are set, and if no explicit configuration is given when constructing the client.
+But clients differ in their use of these variables.
+
+This is problematic because not all code will want to configure settings based on
+environment variables, for example workers that take their configuration from
+configuration files.
+
+This has also caused issues with testing and deploying software, where the
+"hidden" context of environment variables can cause tests to pass or fail, or a
+service to fail when deployed in a different context.
+
+Finally, we should strive for some similarity in usage of the client libraries.
+
+# Details
+
+The proposed change is that client libraries do not *automatically* look to
+these four environment variables for configuration.  The libraries would still have
+options to explicitly "opt in" to using the environment variables, e.g.,
```

Comment:
It seems useful to standardise the meaning of the relevant environment variables.  How they're ready(e.g. constructor option, different constructor) isn't really relevant, but I think it would be good to say that if provided, the variables must be of a specific format, as well as how those are to be interpreted by whichever system reads them.  This way, we can establish the expectation that all tools which use environment variables can use the same environment variables.

Reply:
I think that's a good idea.  I'm conscious of scope-creeping, since I'd like to land this at about the same time as #128, but hopefully this is a minor enough addition that it can "slip right in".